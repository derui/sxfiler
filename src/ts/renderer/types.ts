import * as types from 'sxfiler/common/types';

export interface PaneInfo {
  currentPath: string;
  fileList: types.File[];
  selected: number;
  pane: string;
}

export interface PaneInfos {
  [pane: string]: PaneInfo;
}

// state of global.
export interface GlobalState {
  // state for each pane
  paneInfo: PaneInfos;

  // current located pane
  currentPane: string;
}
