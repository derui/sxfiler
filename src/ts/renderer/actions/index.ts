import * as directoryM from './directory';
import * as paneOperationM from './pane-operations';
import * as windowM from './window';

export const directory = directoryM;

export const paneOperation = paneOperationM;

export const window = windowM;
