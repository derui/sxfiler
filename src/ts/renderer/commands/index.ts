import * as Directory from './defaults/directory';
import * as Pane from './defaults/pane';
import * as Window from './defaults/window';

// export default commands.
import * as types from './types';

// exporting all types from current module.
export * from './types';

export const defaultCommands: types.Commands = {
  commands: [
    Directory.upDirectory,
    Directory.downDirectory,
    Pane.upCursor,
    Pane.downCursor,
    Pane.changePane,
    Window.quitApplication
  ]
};
