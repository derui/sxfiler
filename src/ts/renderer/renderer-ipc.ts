import R from 'ramda';
import Bus from 'sxfiler/renderer/utils/bus';
import {IPC_KEYS} from 'sxfiler/common/constants';
import uuid from 'uuid';

interface Event {
  sender: {
    send: (event: string, ...args: any[]) => void;
  };
}

export interface RendererIPC {
  /**
   * Send a message to channel with args.
   *
   * Notice: When you send message from this, you must take channel that have 'REQUEST_' prefix.
   *
   * Return an observeable sequence that only subscribe event once. You should subscribe it or
   * map/connect/concat... any observeable operations with any other sequence what you want.
   *
   * @param {string} channel channel for sending message
   * @param {...any} args message to send channel
   * @return {Rx.Observerable} An observeable sequence only subscribe once.
   */
  send(channel: string, ...args: any[]): Rx.Observable<any>;

  /**
   * Subscribe observer to channel bus.
   *
   * Using this method if you subscribe all values streaming in specified channel.
   * You do not forget to call dispose from returning disposable.
   *
   * @param {string} channel channel for subscribing observer
   * @param {Rx.Observer} observer observer to subscribe for channel
   * @return {Rx.Disposable} The disposable
   */
  subscribe(channel: string, observer: (val: any) => void): Rx.Disposable;
}

/**
 * IPC event management for Renderer process.
 *
 * This class is based on {Rx.Observable} to publish event handling, so it only
 * provide `subscribe` and `send` method.
 */
export default class RendererIPCImpl implements RendererIPC {

  private _ipc: any;
  private _bus: Bus = new Bus();

  /**
   * Construct RendererIPCHandler
   *
   * Notice for parameter ipc has to require from provided 'require' from 'Electron'.
   *
   * @param {!IPC} ipc ipc module using this
   */
  constructor(ipc: any) {
    this._ipc = ipc;

    this._ipc.on(IPC_KEYS.FINISH_FILES_IN_DIRECTORY, this._onFinishFilesInDirectory.bind(this));
  }

  /**
   * Send a message to channel with args.
   *
   * Notice: When you send message from this, you must take channel that have 'REQUEST_' prefix.
   *
   * Return an observeable sequence that only subscribe event once. You should subscribe it or
   * map/connect/concat... any observeable operations with any other sequence what you want.
   *
   * @param {string} channel channel for sending message
   * @param {...any} args message to send channel
   * @return {Rx.Observerable} An observeable sequence only subscribe once.
   */
  public send(channel: string, ...args: any[]): Rx.Observable<any> {
    let currentId: string = uuid.v4();
    this._ipc.send(channel, ...args.concat([currentId]));

    return this._bus.bus(channel).filter((_args: any[]) => R.last(_args) === currentId).first();
  }

  /**
   * Subscribe observer to channel bus.
   *
   * Using this method if you subscribe all values streaming in specified channel.
   * You do not forget to call dispose from returning disposable.
   *
   * @param {string} channel channel for subscribing observer
   * @param {Rx.Observer} observer observer to subscribe for channel
   * @return {Rx.Disposable} The disposable
   */
  public subscribe(channel: string, observer: (val: any) => void): Rx.Disposable {
    return this._bus.bus(channel).subscribe(observer);
  }

  /**
   * @param {event} ev Event object from ipc event emitter
   * @param {object} err error information for requesting file informations
   * @param {string} path An absolute path what traversed
   * @param {array[object]} files fs.Stats objects in requested directory
   * @param {Pane} pane side of sending request
   * @param {string} uuid The unique event id
   */
  private _onFinishFilesInDirectory(ev: Event, ...args: any[]): void {
    let bus: Rx.Subject<any> = this._bus.bus(IPC_KEYS.REQUEST_FILES_IN_DIRECTORY);

    bus.onNext(args);
  }
}
