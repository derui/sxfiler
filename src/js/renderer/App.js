import React from 'react';
import ReactDOM from 'react-dom';
import Rx from 'rx';

import Main from './components/Main';

import RendererIPC from './RendererIPC';
import {Pane} from 'sxfiler/common/Constants';
import Store from './stores/Store';
import DirectoryBinder from './binders/Directory';
import KeyboardBinder from './binders/Keyboard';
import * as DirectoryAction from './actions/Directory';

let directory = new Store({
  leftPane: {
    path: '.',
    fileList: []
  },
  rightPane: {
    path: '.',
    fileList: []
  }
});

let paneInfo = new Store({
  current: Pane.LEFT,
  left: {
    position: 0
  },
  right: {
    position: 0
  }
});

let binder = new DirectoryBinder();
let actionBinder = new KeyboardBinder();

let state = Rx.Observable.combineLatest(
  directory.getSubject(),
  paneInfo.getSubject(),
  (directoryState, paneInfo) =>
    ({
      directory: directoryState,
      paneInfo
    })
);

state.subscribe((state) => {
  ReactDOM.render(React.createElement(Main, state), document.querySelector('.entry'));
});

let ipc = new RendererIPC(window.require('ipc'));
binder.bind(ipc, directory);
actionBinder.bind(ipc, paneInfo);

DirectoryAction.requestDirectory('.', Pane.LEFT);
DirectoryAction.requestDirectory('.', Pane.RIGHT);
