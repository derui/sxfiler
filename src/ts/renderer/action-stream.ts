import * as Rx from 'rx';
import {EventEmitter} from 'events';
import {RendererIPC} from './renderer-ipc';

type actionArgument = (ipc: RendererIPC) => Rx.Observable<any> | Rx.Observable<any>;

/**
 * An interface of ActionStream
 */
export interface ActionStream {
  /**
   * Publish action to action stream.
   *
   * @param {function(ipc:RendererIPC):Rx.Observable|Rx.Observable} act - an action to publish
   */
  doAction(act: actionArgument): void;

  /**
   * Subscribe events from this.
   *
   * @param {function(value: any)|Rx.Observer<any>} onNext - an onNext callback or observer
   * @param {function(exception: any)} onError - an onError callback
   * @param {function()} onCompleted - an onCompleted callback
   * @return {Rx.IDisposable} a disposable
   */
  subscribe(onNext?: ((value: any) => void|Rx.Observer<any>), onError?: (exception: any) => void, onCompleted?: () => void)
  : Rx.IDisposable;

  /**
   * Dispose published stream from this.
   */
  dispose(): void;
}

/**
 * ActionStream provide action emitter and method to subscribe it.
 */
export default class ActionStreamImpl implements ActionStream {
  private _e: NodeJS.EventEmitter;
  private _broadcastingObservable: Rx.Observable<any>;
  private _broadcastingDisposable: Rx.IDisposable;

  public constructor(ipc: RendererIPC) {
    this._e = new EventEmitter();

    // define observable from EventEmitter
    this._broadcastingObservable = Rx.Observable.fromEventPattern(
      (h: (data: any) => any) => this._e.addListener('action', h),
      (h: (data: any) => any) => this._e.removeListener('action', h),
      (act: any) => {
        if (typeof act === 'function') {
          return (<actionArgument>act)(ipc);
        } else {
          return act;
        }
      }
    ).flatMap((x: Rx.Observable<any>) => x).publish();

    this._broadcastingDisposable = this._broadcastingObservable.connect();
  }

  /**
   * Publish action to action stream.
   *
   * @param {function(ipc:RendererIPC):Rx.Observable|Rx.Observable} act - an action to publish
   */
  public doAction(act: actionArgument): void {
    this._e.emit('action', act);
  }

  /**
   * Subscribe events from this.
   *
   * @param {function(value: any)|Rx.Observer<any>} onNext - an onNext callback or observer
   * @param {function(exception: any)} onError - an onError callback
   * @param {function()} onCompleted - an onCompleted callback
   * @return {Rx.IDisposable} a disposable
   */
  public subscribe(onNext?: ((value: any) => void|Rx.Observer<any>), onError?: (exception: any) => void, onCompleted?: () => void)
  : Rx.IDisposable {
    return this._broadcastingObservable.subscribe(onNext, onError, onCompleted);
  }

  /**
   * Dispose published stream from this.
   */
  public dispose(): void {
    if (this._broadcastingDisposable) {
      this._broadcastingDisposable.dispose();
    }
    this._broadcastingDisposable = null;
  }
}
