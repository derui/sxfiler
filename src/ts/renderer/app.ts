/// <reference path="../../../typings/tsd.d.ts" />
import * as Rx from 'rx';

import * as ReactDOM from 'react-dom';
import * as React from 'react';
import uuid from 'uuid';
import R from 'ramda';

import Main from './components/main';
import ActionStream from './action-stream';
import RendererIPC from './renderer-ipc';
import {PANE} from 'sxfiler/common/constants';
import Directory from './binders/directory';
import CurrentPane from './binders/current-pane';
import * as types from 'sxfiler/renderer/types';
import * as actions from './actions';
import KeyHandler from './key-handlers';
import {convertEventToDescriptor} from 'sxfiler/renderer/utils/key-binding-util';

let ipc = new RendererIPC(window.require('electron').ipcRenderer);
let actionStream = new ActionStream(ipc);

let currentPane = new CurrentPane(PANE.LEFT, actionStream);
let leftPane = new Directory(PANE.LEFT, actionStream);
let rightPane = new Directory(PANE.RIGHT, actionStream);
let paneInfo = Rx.Observable.combineLatest(
  leftPane.store.getObservable(),
  rightPane.store.getObservable(),
  (left: types.PaneInfo, right: types.PaneInfo): types.PaneInfos => {
    return {
      [PANE.LEFT]: left,
      [PANE.RIGHT]: right
    };
  });

let globalState = Rx.Observable.combineLatest(
  paneInfo,
  currentPane.store.getObservable(),
  (paneInfo: types.PaneInfos, currentPane: string): types.GlobalState => {
    return {
      paneInfo,
      currentPane
    };
  }).publish();

globalState.connect();

globalState.subscribe((state: types.GlobalState) => {
  ReactDOM.render(React.createElement(Main, {state, actionStream}), document.querySelector('.entry'));
});

// setup keyevent stream
interface StateWithEvent {
  event: KeyboardEvent;
  eventId: string;
  state: types.GlobalState;
};

let keyEvents = Rx.Observable.combineLatest(
  Rx.Observable.fromEvent(document.body, 'keydown')
    .select((e: KeyboardEvent) => {
      return {
        event: e,
        id: uuid.v4()
      };
    }),
  globalState,
  (ev: {event: KeyboardEvent, id: string}, state: types.GlobalState): StateWithEvent => {
    return {
      event: ev.event,
      eventId: ev.id,
      state
    };
  });

// this variable only store event id from observable sequence as `keyEvents'
let _lastEventId = '';
let keyHandler = new KeyHandler();
keyEvents.subscribe((state: StateWithEvent) => {
  if (state.eventId === _lastEventId) {
    return;
  }

  _lastEventId = state.eventId;

  let _e = R.clone(state.event);
  let desc = convertEventToDescriptor(_e);
  actionStream.doAction(
    keyHandler.invokeCommand(desc, state.state)
  );
});

actionStream.doAction(actions.directory.getDirectory('.', PANE.LEFT));
actionStream.doAction(actions.directory.getDirectory('.', PANE.RIGHT));
