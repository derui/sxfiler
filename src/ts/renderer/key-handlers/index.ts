// provide key handler with key bindings
import * as Rx from 'rx';
import R from 'ramda';
import * as types from 'sxfiler/common/types';
import {GlobalState} from 'sxfiler/renderer/types';
import * as commands from 'sxfiler/renderer/commands';
import {getKeyIdentifier, bindMappingWithCommands} from 'sxfiler/renderer/utils/key-binding-util';
import defaultMapping from './defaults';
import * as actions from 'sxfiler/renderer/actions';
import * as Maybe from 'sxfiler/common/maybe';

let defaultCommands = commands.defaultCommands;

/**
 * Merge two keyDescriptors to one keyDescriptors.
 *
 * @param {array[KeyDescriptor]} src1 key bindings to merge with src2
 * @param {array[KeyDescriptor]} src2 key bindings to merge with src1
 * @return {array[KeyDescriptor]} merged key descriptors
 */
function mergeDescriptors(src1: types.KeyDescriptor[], src2: types.KeyDescriptor[]): types.KeyDescriptor[] {
  'use strict';

  let toObj: any = R.reduce((memo: any, v: types.KeyDescriptor) => R.merge(memo, {[getKeyIdentifier(v)]: v}), {});
  let src1Descs = toObj(src1), src2Descs = toObj(src2);

  return R.values(R.merge(src1Descs, src2Descs));
}

export interface KeyHandler {
  /**
   * Invoke a command binded a key-combination which calculated from KeyboardEvent.
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @param {IGlobalState} globalState - the global state to pass command
   * @return {Rx.Observable} result of command
   */
  invokeCommand(desc: types.KeyDescriptor, globalState: GlobalState): Rx.Observable<any>;
}

/**
 * KeyHandler provide match keyboard input with keybindings that are defined user or default,
 *  and run command in it.
 */
export default class KeyHandlerImpl implements KeyHandler {
  private keyBindings: types.KeyBindings<commands.Command>;

  /**
   * Construct KeyHandler.
   *
   * @param {IKeyBindings} [keyBindings] key bindings to append to default binding
   */
  constructor(keyBindings?: types.KeyBindings<commands.Command>) {
    this.keyBindings = bindMappingWithCommands(defaultMapping, defaultCommands);

    // if keybindings given, merge with default bindings.
    if (keyBindings) {
      this.keyBindings.bindings = mergeDescriptors(this.keyBindings.bindings, keyBindings.bindings);
      this.keyBindings.mapping = R.merge(this.keyBindings.bindings, keyBindings.mapping);
    }
  }

  /**
   * Invoke a command binded a key-combination which calculated from KeyboardEvent.
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @param {IGlobalState} globalState - the global state to pass command
   * @return {Rx.Observable} result of command
   */
  public invokeCommand(desc: types.KeyDescriptor, globalState: GlobalState): Rx.Observable<any> {

    let command = this.matchKey(desc);
    if (Maybe.isNone(command)) {
      return Rx.Observable.empty();
    }

    return command.val.command(globalState, actions);
  }

  /**
   * Match given KeyboardEvent with keybindings in this class
   *
   * @param {KeyDescriptor} desc - a descriptor to invoke command
   * @return {ICommand} a command if match keyboard event with keybindings. return null if not.
   */
  private matchKey(desc: types.KeyDescriptor): Maybe.Data<commands.Command> {
    if (!desc) {
      return Maybe.NONE;
    }

    let id = getKeyIdentifier(desc);

    let binding = this.keyBindings.mapping[id];
    if (!binding) {
      return Maybe.NONE;
    }
    return Maybe.just(binding);
  }
}
