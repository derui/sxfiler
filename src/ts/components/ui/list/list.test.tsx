import * as React from "react";
import renderer from "react-test-renderer";

import List from "./list";

describe("UI kit", () => {
  describe("List Item", () => {
    it("render with empty nodes", () => {
      const tree = renderer.create(<List />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("render with single nodes", () => {
      const tree = renderer.create(<List>test</List>).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("can render with other container", () => {
      const tree = renderer.create(<List container="span">test</List>).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("render with single nodes", () => {
      const tree = renderer
        .create(
          <List>
            <span className="foo">foo</span>
            <a href="">link</a>
          </List>
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("can pass names of class to apply style", () => {
      const tree = renderer
        .create(
          <List classes={["a", "b"]}>
            <span className="foo">foo</span>
          </List>
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("can pass props directly", () => {
      const onClick = () => null;
      const tree = renderer
        .create(
          <List classes={["a", "b"]} id="foo" onClick={onClick}>
            <span className="foo">foo</span>
          </List>
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
