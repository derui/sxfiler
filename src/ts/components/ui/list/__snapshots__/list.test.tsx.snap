// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI kit List Item can pass names of class to apply style 1`] = `
<div
  className="a b"
  role="list"
>
  <span
    className="foo"
  >
    foo
  </span>
</div>
`;

exports[`UI kit List Item can pass props directly 1`] = `
<div
  className="a b"
  id="foo"
  onClick={[Function]}
  role="list"
>
  <span
    className="foo"
  >
    foo
  </span>
</div>
`;

exports[`UI kit List Item can render with other container 1`] = `
<span
  role="list"
>
  test
</span>
`;

exports[`UI kit List Item render with empty nodes 1`] = `
<div
  role="list"
/>
`;

exports[`UI kit List Item render with single nodes 1`] = `
<div
  role="list"
>
  test
</div>
`;

exports[`UI kit List Item render with single nodes 2`] = `
<div
  role="list"
>
  <span
    className="foo"
  >
    foo
  </span>
  <a
    href=""
  >
    link
  </a>
</div>
`;
